### DISCUSSION TITLE
Adapt LoopSwapTrans to allow more complex loop reordering

### DISCUSSION THREAD

**Discussion by LonelyCat124** [stfc/PSyclone#2715] [Ideas]
After Mark's talk today, I think we have no inbuilt transformation to enable loop reordering other than for the basic case of `LoopSwapTrans`.

So for example if we had a case where we have 5 nested loops and wanted to move the inner-most loop up two levels we'd have to essentially call LoopSwapTrans twice on its parent , which would be a bit complicated to generalise for a code-base (e.g. if we wanted to automatically move the k loop in LFRic to be the innermost collapsable loop whenever possible for GPUs).

We'd probably also want some sort of correctness checking if the reordered loop is safe for some level of parallelism, but I couldn't construct a failing example right now.

I'm not sure how easy this would be (and thus whether its worthwhile over Fortran code refactoring) but it would depend on if its likely to be useful in the original source codes.

**Comment by arporter** [comment]
We have `HoistTrans` but that currently only works with `Assignment` nodes. Perhaps we should generalise this?
